### ISSUE TITLE
BUG: Eyrie is missing a turn in this game

### ISSUE THREAD

**Issue by AmasaDelano** [Vagabottos/RootVis#36]
https://rootvis.seiyria.com/full-game?gameUrl=https:%2F%2Fraw.githubusercontent.com%2FVagabottos%2FRootlog%2Fmaster%2FGames%2F2020_11_19_winter_tournament_r1g2.rootlog&clearing=4&action=176

Hit "Next Action" to watch the Eyrie's turn get skipped.

The parser seems to parse correctly when I test it with this game locally. I'm not sure where the issue is, recording it here.

**Comment by AmasaDelano** [comment]
Update from debugging: The turn isn't missing, it's just out of order. [Here](https://rootvis.seiyria.com/full-game?gameUrl=https:%2F%2Fraw.githubusercontent.com%2FVagabottos%2FRootlog%2Fmaster%2FGames%2F2020_11_19_winter_tournament_r1g2.rootlog&clearing=4&action=212)'s the missing turn.

There's also a second "missing" turn (for the Vagabond this time) right [here](https://rootvis.seiyria.com/full-game?gameUrl=https:%2F%2Fraw.githubusercontent.com%2FVagabottos%2FRootlog%2Fmaster%2FGames%2F2020_11_19_winter_tournament_r1g2.rootlog&clearing=4&action=333).

**Comment by AmasaDelano** [comment]
The issue seems to be in the `getAllActions` function in `rootlog.service.ts`. `game.turns` is set correctly then, but `allActions` comes out with the actions out of order.

**Comment by AmasaDelano** [comment]
The issue seems to be that the call to `.flat` is rearranging the actions within the game. I'm not sure why. I've tried several solutions so far, with no luck.

Here's stuff that didn't work (after renaming the current `allActions` assignment to `allTurns`):

`const allActions = Array.prototype.concat.apply([], allTurns);`

```
allTurns.forEach(turn => {
  turn.forEach(action => allActions.push(action));
});
```

After running either of the above, the actions in `allActions` were out of order.

If anyone else has any insight on this, that'd be great. I'm pretty stumped.

**Comment by seiyria** [comment]
There isn't a logistical reason that flat would rearrange anything.

```
[
[1,2],
[3,4]
].flat() === [1,2,3,4]
```

I can't offer any insight beyond that.

**Comment by AmasaDelano** [comment]
Thank you! I know there isn't any reason it would, that's why I'm stumped at the moment.

**Comment by chadfraser** [comment]
Has this been fixed? On my end, I don't see a missing turn with that first link: It goes from `(177/375)[Vagabond][[needs move description]] #->V` to `(178/375)[Eyrie][[needs move description]] 2F#E->$_r` as expected

**Comment by AmasaDelano** [comment]
> Has this been fixed? On my end, I don't see a missing turn with that first link: It goes from `(177/375)[Vagabond][[needs move description]] #->V` to `(178/375)[Eyrie][[needs move description]] 2F#E->$_r` as expected

Have you force-refreshed to make sure you're pulling the latest version? This issue is still happening for me.

On the bright side, if force-refreshing _does_ cause this issue to happen for you, that means that a fairly recent change introduced this issue, which should make it easier to track down. :)

**Comment by AmasaDelano** [comment]
Funnily enough, solved by #37. The real issue here (besides the missing turn) is that the game that was linked had `parseErrors` on it, but it was being loaded anyway. The game as is should never have been loaded.

I've corrected the game, which should fix that, but I've also made entries in `parseErrors` mark the game as "invalid," so this sort of thing shouldn't happen again.

Thanks! :)
